### DISCUSSION TITLE
How to learn these cool technologies?

### DISCUSSION THREAD

**Discussion by bharadwajpro** [ggml-org/llama.cpp#1244] [General]
I find the development being done here very interesting. I am familiar with ML being done using Python only. But there comes a point where you have to think about optimizing the models/algorithms on consumer devices. Feel free to close this issue but I want to know from the author and contributors, how to learn the math optimizations done on different CPU/GPU architectures, how to find these APIs and the different ML algorithms being implemented? Any beginner friendly resources or courses or repositories?

Thank you.

**Comment by gotzmann** [comment]
Just learn C/C++ (and maybe assembler) and after some years you'll catch up those things :)
